src/homestead_xml_utils.cpp

```diff
@@ -24,17 +24,47 @@ namespace XmlUtils
 
 // Builds a ClearwaterRegData XML document for passing to Sprout,
 // based on the given registration state and User-Data XML from the HSS.
+// Will also include a PreviousRegistrationState element if a known
+// previous registration state is provided.
 int build_ClearwaterRegData_xml(ImplicitRegistrationSet* irs,
                                 std::string& xml_str)
 {
+  return build_ClearwaterRegData_xml(irs,
+                                     xml_str,
+                                     RegistrationState::UNKNOWN);
+}
+
+// As above, but will also include a PreviousRegistrationState element if a known
+// previous registration state is provided.
+int build_ClearwaterRegData_xml(ImplicitRegistrationSet* irs,
+                                std::string& xml_str,
+                                RegistrationState prev_reg_state)
+{
   rapidxml::xml_document<> doc;
   rapidxml::xml_node<>* root = doc.allocate_node(rapidxml::node_type::node_element,
                                                  RegDataXMLUtils::CLEARWATER_REG_DATA);
 
-  // Add the registration state - note we need to pass in the regtype string
-  // here to ensure it isn't destroyed before the XML is printed.
-  std::string regtype;
-  add_reg_state_node(irs->get_reg_state(), doc, root, regtype);
+  // Add the registration state - note we need to pass in a string that can be
+  // used to store the registration state and whose life time is the same as
+  // that of the XML doc.
+  std::string state_string;
+  add_reg_state_node(irs->get_reg_state(),
+                     doc,
+                     root,
+                     RegDataXMLUtils::REGISTRATION_STATE,
+                     state_string);
+
+  // If we have been provided with a prev_reg_state then add a
+  // prev_reg_state node.
+  std::string previous_statestring;
+  if (prev_reg_state != RegistrationState::UNKNOWN)
+  {
+    add_reg_state_node(prev_reg_state,
+                       doc,
+                       root,
+                       RegDataXMLUtils::prev_reg_state,
```

Comment:
This looks very inconsistent compared to line 54

Reply:
Argh.   Bad check-in.   It's not inconsistent -- it's just completely broken!